runtime/compiler/control/CompilationThread.cpp

```diff
@@ -10462,71 +10456,132 @@ TR::CompilationInfo::increaseUnstoredBytes(U_32 aotBytes, U_32 jitBytes)
 #endif
    }
 
+
+// Compute free physical memory taking into account container limits
+// To limit overhead, this function caches the computed value and
+// only refreshes it if more than 0.5 sec have passed since the last update
+// or if "now" parameter is true
+// Can return OMRPORT_MEMINFO_NOT_AVAILABLE as error code
+// incompleteInfo flag is set to indicate that buffered/cached memory could not be read
+// Synchronization issues?
 uint64_t
-TR::CompilationInfo::computeFreePhysicalMemory(bool &incompleteInfo)
+TR::CompilationInfo::computeFreePhysicalMemory(bool &incompleteInfo, J9JITConfig *jitConfig)
    {
    bool incomplete = false;
+   PORT_ACCESS_FROM_JITCONFIG(jitConfig);
 
-   if (_cachedFreePhysicalMemoryB >= 0)
+   uint64_t freePhysicalMemory = OMRPORT_MEMINFO_NOT_AVAILABLE;
+   J9MemoryInfo memInfo;
+   if (0 == j9sysinfo_get_memory_info(&memInfo)
+      && memInfo.availPhysical != OMRPORT_MEMINFO_NOT_AVAILABLE
+      && memInfo.hostAvailPhysical != OMRPORT_MEMINFO_NOT_AVAILABLE)
       {
-      static int64_t lastUpdateTime = 0;
-      int64_t crtElapsedTime = getPersistentInfo()->getElapsedTime();
+      freePhysicalMemory = memInfo.availPhysical;
+      uint64_t freeHostPhysicalMemorySizeB = memInfo.hostAvailPhysical;
 
-      // time to recompute freePhysicalMemory
-      if (lastUpdateTime == 0
-          || (crtElapsedTime - lastUpdateTime) > (int64_t)TR::Options::getUpdateFreeMemoryMinPeriod())
-         {
-         PORT_ACCESS_FROM_JITCONFIG(_jitConfig);
+      // TODO: do we have this as unavailable in docker?
+      if (memInfo.cached != OMRPORT_MEMINFO_NOT_AVAILABLE) // Note 'cached' is not available in containers
+         freePhysicalMemory += memInfo.cached;
+      else
+         incomplete = true;
 
-         J9MemoryInfo memInfo;
-         if (0 == j9sysinfo_get_memory_info(&memInfo)
-             && memInfo.availPhysical != OMRPORT_MEMINFO_NOT_AVAILABLE)
-            {
-            uint64_t freePhysicalMemorySizeB = memInfo.availPhysical;
-           
-            if (memInfo.cached != OMRPORT_MEMINFO_NOT_AVAILABLE)
-               freePhysicalMemorySizeB += memInfo.cached;
-            else
-               incomplete = true;
+      if (memInfo.hostCached != OMRPORT_MEMINFO_NOT_AVAILABLE) // Note 'cached' is not available in containers
+         freeHostPhysicalMemorySizeB += memInfo.hostCached;
+      else
+         incomplete = true;
 #if defined(LINUX)
-            if (memInfo.buffered != OMRPORT_MEMINFO_NOT_AVAILABLE)
-               freePhysicalMemorySizeB += memInfo.buffered;
-            else
-               incomplete = true;
+      if (memInfo.buffered != OMRPORT_MEMINFO_NOT_AVAILABLE)
```

Comment:
As per https://github.com/eclipse/omr/blob/4e352d639ed726de60b827a0e922be5181180729/port/unix/omrsysinfo.c#L1616 `buffered` is not available in containers. Does that mean we would always set `incomplete` to `true` when running in containers?

Reply:
Thinking about this, we don't want to set incomplete for container, thus there are two possible solutions
1) In `retrieveLinuxCgroupMemoryStats` always fill with 0 the buffered field instead of UNAVAILABLE
2) In `computeFreePhysicalMemory` check whether we are in a container by using
`if (OMR_CGROUP_SUBSYSTEM_MEMORY == omrsysinfo_cgroup_are_subsystems_enabled(OMR_CGROUP_SUBSYSTEM_MEMORY)) `
I might go with solution (2) because I think I should not add the cache value to the free physical memory. The reason is that on the host the OS can increase the physical memory by releasing some cached pages, but in container the available memory is (limit-used) and any cached pages are not going to help.